File: src/main/java/com/reactive/nexo/siau/service/PqrsdfService.java
package com.reactive.nexo.siau.service;

import com.reactive.nexo.siau.dto.*;
import com.reactive.nexo.siau.entity.*;
import com.reactive.nexo.siau.model.*;
import com.reactive.nexo.siau.repository.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import reactor.core.scheduler.Schedulers;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class PqrsdfService {

    private static final Logger log = LoggerFactory.getLogger(PqrsdfService.class);

    private final PqrsdfRepository pqrsdfRepo;
    private final AsignacionRepository asignacionRepo;
    private final TrazabilidadRepository trazabilidadRepo;
    private final DepartamentoRepository departamentoRepo;

    public PqrsdfService(PqrsdfRepository pqrsdfRepo,
                         AsignacionRepository asignacionRepo,
                         TrazabilidadRepository trazabilidadRepo,
                         DepartamentoRepository departamentoRepo) {
        this.pqrsdfRepo = pqrsdfRepo;
        this.asignacionRepo = asignacionRepo;
        this.trazabilidadRepo = trazabilidadRepo;
        this.departamentoRepo = departamentoRepo;
    }

    public PqrsdfResponse create(CreatePqrsdfRequest req, Long employeeId) {
        // Generar consecutivo automático
        Long maxId = pqrsdfRepo.findMaxId()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(0L);
        String consecutivo = generarConsecutivo(maxId + 1, req.getTipo());

        LocalDate hoy = LocalDate.now();
        String hora = req.getHoraRadicado() != null ? req.getHoraRadicado()
                : LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm"));

        PqrsdfEntity entity = PqrsdfEntity.builder()
                .consecutivo(consecutivo)
                .tipo(req.getTipo())
                .fechaRadicado(req.getFechaRadicado() != null ? req.getFechaRadicado() : hoy)
                .horaRadicado(hora)
                .nombresUsuario(req.getNombresUsuario())
                .tipoDocumento(req.getTipoDocumento())
                .numeroDocumento(req.getNumeroDocumento())
                .telefono(req.getTelefono())
                .email(req.getEmail())
                .direccion(req.getDireccion())
                .eps(req.getEps())
                .regimen(req.getRegimen())
                .medioRecepcion(req.getMedioRecepcion())
                .servicioInvolucrado(req.getServicioInvolucrado())
                .funcionarioInvolucrado(req.getFuncionarioInvolucrado())
                .descripcion(req.getDescripcion())
                .clasificacion(req.getClasificacion())
                .estado(EstadoPqrsdf.RECIBIDO.getValue())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .createdBy(req.getCreatedBy() != null ? req.getCreatedBy() : employeeId)
                .actaBuzonId(req.getActaBuzonId())
                .build();

        PqrsdfEntity saved = pqrsdfRepo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .block();

        // Registrar trazabilidad
        registrarTrazabilidad(saved.getId(), "CREACION",
                "PQRSDF " + consecutivo + " creada - Tipo: " + req.getTipo(),
                employeeId, null);

        log.info("PQRSDF {} creada: {}", consecutivo, saved.getId());
        return mapToResponse(saved, null, null, 0);
    }

    public Optional<PqrsdfResponse> findById(Long id) {
        PqrsdfEntity entity = pqrsdfRepo.findById(id)
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (entity == null) return Optional.empty();

        // Obtener última asignación si existe
        AsignacionEntity ultimaAsignacion = asignacionRepo.findUltimaByPqrsdfId(id)
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(null);

        String deptoNombre = null;
        String estadoAsig = null;
        if (ultimaAsignacion != null) {
            estadoAsig = ultimaAsignacion.getEstado();
            DepartamentoEntity depto = departamentoRepo.findById(ultimaAsignacion.getDepartamentoId())
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(null);
            if (depto != null) deptoNombre = depto.getNombre();
        }

        int diasTranscurridos = (int) ChronoUnit.DAYS.between(entity.getFechaRadicado(), LocalDate.now());
        return Optional.of(mapToResponse(entity, deptoNombre, estadoAsig, diasTranscurridos));
    }

    public List<PqrsdfResponse> search(Map<String, String> filters, int page, int size) {
        List<PqrsdfEntity> all = pqrsdfRepo.findAll()
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());

        // Aplicar filtros en memoria (por simplicidad, igual que el patrón de appointments)
        if (filters.containsKey("estado")) {
            String est = filters.get("estado");
            all = all.stream().filter(e -> e.getEstado().equalsIgnoreCase(est)).collect(Collectors.toList());
        }
        if (filters.containsKey("tipo")) {
            String tip = filters.get("tipo");
            all = all.stream().filter(e -> e.getTipo().equalsIgnoreCase(tip)).collect(Collectors.toList());
        }
        if (filters.containsKey("numeroDocumento")) {
            String doc = filters.get("numeroDocumento");
            all = all.stream().filter(e -> doc.equals(e.getNumeroDocumento())).collect(Collectors.toList());
        }
        if (filters.containsKey("fechaDesde")) {
            LocalDate desde = LocalDate.parse(filters.get("fechaDesde"));
            all = all.stream().filter(e -> !e.getFechaRadicado().isBefore(desde)).collect(Collectors.toList());
        }
        if (filters.containsKey("fechaHasta")) {
            LocalDate hasta = LocalDate.parse(filters.get("fechaHasta"));
            all = all.stream().filter(e -> !e.getFechaRadicado().isAfter(hasta)).collect(Collectors.toList());
        }

        // Ordenar por fecha descendente
        all.sort((a, b) -> b.getCreatedAt().compareTo(a.getCreatedAt()));

        int from = Math.min(page * size, all.size());
        int to = Math.min(from + size, all.size());
        List<PqrsdfEntity> pageList = all.subList(from, to);

        List<PqrsdfResponse> result = new ArrayList<>();
        for (PqrsdfEntity e : pageList) {
            int dias = (int) ChronoUnit.DAYS.between(e.getFechaRadicado(), LocalDate.now());
            result.add(mapToResponse(e, null, null, dias));
        }
        return result;
    }

    public long count(Map<String, String> filters) {
        if (filters.isEmpty()) {
            return pqrsdfRepo.countTotal()
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(0L);
        }
        // Si hay filtros, contar desde la lista completa
        List<PqrsdfEntity> all = pqrsdfRepo.findAll()
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());
        return all.size();
    }

    public PqrsdfResponse update(Long id, UpdatePqrsdfRequest req, Long employeeId) {
        PqrsdfEntity entity = pqrsdfRepo.findById(id)
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (entity == null) return null;

        if (req.getClasificacion() != null) entity.setClasificacion(req.getClasificacion());
        if (req.getObservaciones() != null) entity.setObservaciones(req.getObservaciones());
        entity.setUpdatedAt(LocalDateTime.now());
        entity.setUpdatedBy(req.getUpdatedBy() != null ? req.getUpdatedBy() : employeeId);

        PqrsdfEntity saved = pqrsdfRepo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .block();

        registrarTrazabilidad(id, "ACTUALIZACION",
                "PQRSDF " + entity.getConsecutivo() + " actualizada",
                employeeId, null);

        return mapToResponse(saved, null, null, 0);
    }

    public PqrsdfResponse responder(Long id, ResponderPqrsdfRequest req, Long employeeId) {
        PqrsdfEntity entity = pqrsdfRepo.findById(id)
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (entity == null) return null;

        entity.setRespuestaFinal(req.getRespuestaFinal());
        entity.setMedioRespuesta(req.getMedioRespuesta());
        entity.setFechaRespuesta(LocalDate.now());
        entity.setEstado(EstadoPqrsdf.RESPONDIDO.getValue());
        if (req.getClasificacion() != null) entity.setClasificacion(req.getClasificacion());
        entity.setUpdatedAt(LocalDateTime.now());
        entity.setUpdatedBy(req.getUpdatedBy() != null ? req.getUpdatedBy() : employeeId);

        PqrsdfEntity saved = pqrsdfRepo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .block();

        // Actualizar asignaciones relacionadas
        List<AsignacionEntity> asignaciones = asignacionRepo.findByPqrsdfId(id)
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());
        for (AsignacionEntity asig : asignaciones) {
            if (!"RESPONDIDA".equals(asig.getEstado())) {
                asig.setEstado("RESPONDIDA");
                asig.setFechaRespuestaArea(LocalDateTime.now());
                asig.setUpdatedAt(LocalDateTime.now());
                asignacionRepo.save(asig)
                        .subscribeOn(Schedulers.boundedElastic())
                        .block();
            }
        }

        registrarTrazabilidad(id, "RESPUESTA",
                "PQRSDF " + entity.getConsecutivo() + " respondida - Medio: " + req.getMedioRespuesta(),
                employeeId, null);

        return mapToResponse(saved, null, null, 0);
    }

    public PqrsdfResponse cerrar(Long id, Long employeeId) {
        PqrsdfEntity entity = pqrsdfRepo.findById(id)
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (entity == null) return null;

        entity.setEstado(EstadoPqrsdf.CERRADO.getValue());
        entity.setUpdatedAt(LocalDateTime.now());
        entity.setUpdatedBy(employeeId);

        PqrsdfEntity saved = pqrsdfRepo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .block();

        registrarTrazabilidad(id, "CIERRE",
                "PQRSDF " + entity.getConsecutivo() + " cerrada",
                employeeId, null);

        return mapToResponse(saved, null, null, 0);
    }

    public List<TrazabilidadResponse> getTrazabilidad(Long pqrsdfId) {
        List<TrazabilidadEntity> list = trazabilidadRepo.findHistorialByPqrsdfId(pqrsdfId)
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());
        List<TrazabilidadResponse> result = new ArrayList<>();
        for (TrazabilidadEntity te : list) {
            result.add(mapTrazabilidad(te));
        }
        return result;
    }

    // --- Métodos auxiliares ---

    private String generarConsecutivo(Long id, String tipo) {
        String anio = String.valueOf(LocalDate.now().getYear());
        return String.format("PQRS-%s-%s-%04d", tipo.toUpperCase(), anio, id);
    }

    private void registrarTrazabilidad(Long pqrsdfId, String accion, String descripcion,
                                       Long usuarioId, String metadataJson) {
        TrazabilidadEntity entity = TrazabilidadEntity.builder()
                .pqrsdfId(pqrsdfId)
                .accion(accion)
                .descripcion(descripcion)
                .usuarioId(usuarioId)
                .createdAt(LocalDateTime.now())
                .metadataJson(metadataJson)
                .build();
        trazabilidadRepo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .subscribe();
    }

    private PqrsdfResponse mapToResponse(PqrsdfEntity e, String ultimoDepto,
                                          String ultimoEstadoAsig, int diasTranscurridos) {
        TipoPqrsdf tipo = TipoPqrsdf.fromCodigo(e.getTipo());
        return PqrsdfResponse.builder()
                .id(e.getId())
                .consecutivo(e.getConsecutivo())
                .tipo(e.getTipo())
                .tipoNombre(tipo.getNombre())
                .fechaRadicado(e.getFechaRadicado())
                .horaRadicado(e.getHoraRadicado())
                .nombresUsuario(e.getNombresUsuario())
                .tipoDocumento(e.getTipoDocumento())
                .numeroDocumento(e.getNumeroDocumento())
                .telefono(e.getTelefono())
                .email(e.getEmail())
                .direccion(e.getDireccion())
                .eps(e.getEps())
                .regimen(e.getRegimen())
                .medioRecepcion(e.getMedioRecepcion())
                .servicioInvolucrado(e.getServicioInvolucrado())
                .funcionarioInvolucrado(e.getFuncionarioInvolucrado())
                .descripcion(e.getDescripcion())
                .clasificacion(e.getClasificacion())
                .estado(e.getEstado())
                .fechaRespuesta(e.getFechaRespuesta())
                .medioRespuesta(e.getMedioRespuesta())
                .respuestaFinal(e.getRespuestaFinal())
                .observaciones(e.getObservaciones())
                .createdAt(e.getCreatedAt())
                .updatedAt(e.getUpdatedAt())
                .createdBy(e.getCreatedBy())
                .updatedBy(e.getUpdatedBy())
                .ultimoDepartamentoAsignado(ultimoDepto)
                .ultimoEstadoAsignacion(ultimoEstadoAsig)
                .diasTranscurridos(diasTranscurridos)
                .build();
    }

    private TrazabilidadResponse mapTrazabilidad(TrazabilidadEntity te) {
        return TrazabilidadResponse.builder()
                .id(te.getId())
                .pqrsdfId(te.getPqrsdfId())
                .accion(te.getAccion())
                .descripcion(te.getDescripcion())
                .usuarioId(te.getUsuarioId())
                .usuarioNombre(te.getUsuarioNombre())
                .metadataJson(te.getMetadataJson())
                .createdAt(te.getCreatedAt())
                .build();
    }
}

File: app/services/pqrsdf_service.py
"""Service layer for PQRSDF entity — creation, search, response, closure, and history tracking."""

from __future__ import annotations

from datetime import date, datetime, timezone
from typing import Any

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func, text
from sqlalchemy.orm import selectinload

from app.repositories import pqrsdf_repo, trazabilidad_repo, asignacion_repo
from app.schemas import (
    CreatePqrsdfRequest,
    UpdatePqrsdfRequest,
    ResponderPqrsdfRequest,
    PagedResponse,
    PqrsdfResponse,
    TrazabilidadResponse,
)
from app.models import Pqrsdf, Asignacion, Trazabilidad
from app.services.catalogos import get_tipo_nombre


# ──────────────────────────────────────────────
# Helpers
# ──────────────────────────────────────────────

def _generar_consecutivo(tipo: str, next_id: int) -> str:
    """Genera consecutivo con formato PQRS-{tipo}-{anio}-{id:04d}."""
    anio = date.today().year
    return f"PQRS-{tipo}-{anio}-{next_id:04d}"


async def _registrar_trazabilidad(
    session: AsyncSession,
    pqrsdf_id: int,
    accion: str,
    descripcion: str,
    usuario_id: int | None = None,
    usuario_nombre: str | None = None,
    metadata_json: str | None = None,
) -> Trazabilidad:
    """Registra una entrada de trazabilidad para una PQRSDF."""
    trazabilidad = Trazabilidad(
        pqrsdf_id=pqrsdf_id,
        accion=accion,
        descripcion=descripcion,
        usuario_id=usuario_id,
        usuario_nombre=usuario_nombre,
        metadata_json=metadata_json,
        created_at=datetime.now(timezone.utc),
    )
    return await trazabilidad_repo.save(session, trazabilidad)


def _map_to_response(
    pqrsdf: Pqrsdf,
    ultimo_depto: str | None = None,
    ultimo_estado_asig: str | None = None,
    dias: int = 0,
) -> PqrsdfResponse:
    """Mapea una entidad Pqrsdf a PqrsdfResponse con datos de asignación."""
    return PqrsdfResponse(
        id=pqrsdf.id,
        consecutivo=pqrsdf.consecutivo,
        tipo=pqrsdf.tipo,
        tipoNombre=get_tipo_nombre(pqrsdf.tipo),
        fechaRadicado=pqrsdf.fecha_radicado if isinstance(pqrsdf.fecha_radicado, date) else None,
        horaRadicado=pqrsdf.hora_radicado,
        nombresUsuario=pqrsdf.nombres_usuario,
        tipoDocumento=pqrsdf.tipo_documento,
        numeroDocumento=pqrsdf.numero_documento,
        telefono=pqrsdf.telefono,
        email=pqrsdf.email,
        direccion=pqrsdf.direccion,
        eps=pqrsdf.eps,
        regimen=pqrsdf.regimen,
        medioRecepcion=pqrsdf.medio_recepcion,
        servicioInvolucrado=pqrsdf.servicio_involucrado,
        funcionarioInvolucrado=pqrsdf.funcionario_involucrado,
        descripcion=pqrsdf.descripcion,
        clasificacion=pqrsdf.clasificacion,
        estado=pqrsdf.estado,
        fechaRespuesta=pqrsdf.fecha_respuesta if isinstance(pqrsdf.fecha_respuesta, date) else None,
        medioRespuesta=pqrsdf.medio_respuesta,
        respuestaFinal=pqrsdf.respuesta_final,
        observaciones=pqrsdf.observaciones,
        createdAt=pqrsdf.created_at,
        updatedAt=pqrsdf.updated_at,
        createdBy=pqrsdf.created_by,
        updatedBy=pqrsdf.updated_by,
        ultimoDepartamentoAsignado=ultimo_depto,
        ultimoEstadoAsignacion=ultimo_estado_asig,
        diasTranscurridos=dias,
    )


async def _get_ultima_asignacion_info(
    session: AsyncSession, pqrsdf_id: int
) -> tuple[str | None, str | None, int]:
    """Obtiene el último departamento, estado de asignación y días transcurridos."""
    stmt = (
        select(Asignacion)
        .where(Asignacion.pqrsdf_id == pqrsdf_id)
        .order_by(Asignacion.created_at.desc())
        .limit(1)
    )
    result = await session.execute(stmt)
    asignacion = result.scalar_one_or_none()

    if asignacion is None:
        return None, None, 0

    dias = 0
    if (
        asignacion.fecha_asignacion
        and asignacion.estado
        and asignacion.estado not in ("RESPONDIDA", "CERRADA")
    ):
        delta = date.today() - asignacion.fecha_asignacion.date()
        dias = delta.days

    depto_nombre = None
    if asignacion.departamento_rel:
        depto_nombre = asignacion.departamento_rel.nombre

    return depto_nombre, asignacion.estado, dias


# ──────────────────────────────────────────────
# CRUD principal
# ──────────────────────────────────────────────

async def create(
    session: AsyncSession,
    req: CreatePqrsdfRequest,
    employee_id: int,
) -> PqrsdfResponse:
    """Crea una nueva PQRSDF con estado RECIBIDO y registra trazabilidad."""
    # Obtener el próximo ID para el consecutivo
    result = await session.execute(select(func.max(Pqrsdf.id)))
    max_id = result.scalar() or 0
    next_id = max_id + 1

    consecutivo = _generar_consecutivo(req.tipo, next_id)

    entity = Pqrsdf(
        consecutivo=consecutivo,
        tipo=req.tipo,
        fecha_radicado=req.fechaRadicado or date.today(),
        hora_radicado=req.horaRadicado
        or datetime.now(timezone.utc).strftime("%H:%M"),
        nombres_usuario=req.nombresUsuario,
        tipo_documento=req.tipoDocumento,
        numero_documento=req.numeroDocumento,
        telefono=req.telefono,
        email=req.email,
        direccion=req.direccion,
        eps=req.eps,
        regimen=req.regimen,
        medio_recepcion=req.medioRecepcion,
        servicio_involucrado=req.servicioInvolucrado,
        funcionario_involucrado=req.funcionarioInvolucrado,
        descripcion=req.descripcion,
        clasificacion=req.clasificacion,
        estado="RECIBIDO",
        created_at=datetime.now(timezone.utc),
        updated_at=datetime.now(timezone.utc),
        created_by=employee_id,
        updated_by=employee_id,
        acta_buzon_id=req.actaBuzonId,
    )

    saved = await pqrsdf_repo.save(session, entity)

    await _registrar_trazabilidad(
        session,
        saved.id,
        "CREACION",
        f"PQRSDF creada con consecutivo {consecutivo}",
        employee_id,
    )

    tipo_nombre = get_tipo_nombre(req.tipo)
    return _map_to_response(saved, ultimo_depto=None, ultimo_estado_asig=None, dias=0)


async def get_by_id(
    session: AsyncSession,
    id: int,
) -> PqrsdfResponse | None:
    """Alias router-compatible para find_by_id."""
    return await find_by_id(session, id)


async def find_by_id(
    session: AsyncSession,
    id: int,
) -> PqrsdfResponse | None:
    """Busca una PQRSDF por ID con información de última asignación."""
    entity = await pqrsdf_repo.find_by_id(session, id)
    if entity is None:
        return None

    depto, estado_asig, dias = await _get_ultima_asignacion_info(session, id)
    return _map_to_response(entity, depto, estado_asig, dias)


async def list_all(
    session: AsyncSession,
    page: int = 0,
    size: int = 10,
    estado: str | None = None,
    tipo: str | None = None,
    numero_documento: str | None = None,
    fecha_desde: str | None = None,
    fecha_hasta: str | None = None,
) -> PagedResponse:
    """Lista PQRSDF con filtros opcionales y paginación (wrapper router-compatible)."""
    filters = {}
    if estado:
        filters["estado"] = estado
    if tipo:
        filters["tipo"] = tipo
    if numero_documento:
        filters["numeroDocumento"] = numero_documento
    if fecha_desde:
        filters["fechaDesde"] = fecha_desde
    if fecha_hasta:
        filters["fechaHasta"] = fecha_hasta
    items = await search(session, filters, page, size)
    total = await count(session, filters)
    return PagedResponse(
        content=items,
        page=page,
        size=size,
        totalElements=total,
        totalPages=max(1, (total + size - 1) // size),
        last=(page + 1) * size >= total,
    )


async def search(
    session: AsyncSession,
    filters: dict[str, Any],
    page: int = 0,
    size: int = 10,
) -> list[PqrsdfResponse]:
    """Busca PQRSDF con filtros, paginación y orden descendente por created_at."""
    query = select(Pqrsdf)

    if filters.get("estado"):
        query = query.where(Pqrsdf.estado == filters["estado"])
    if filters.get("tipo"):
        query = query.where(Pqrsdf.tipo == filters["tipo"])
    if filters.get("numeroDocumento"):
        query = query.where(Pqrsdf.numero_documento == filters["numeroDocumento"])
    if filters.get("fechaDesde"):
        query = query.where(Pqrsdf.fecha_radicado >= filters["fechaDesde"])
    if filters.get("fechaHasta"):
        query = query.where(Pqrsdf.fecha_radicado <= filters["fechaHasta"])
    if filters.get("consecutivo"):
        query = query.where(Pqrsdf.consecutivo.ilike(f'%{filters["consecutivo"]}%'))

    query = query.order_by(Pqrsdf.created_at.desc())
    query = query.offset(page * size).limit(size)

    result = await session.execute(query)
    entities = result.scalars().all()

    responses = []
    for entity in entities:
        depto, estado_asig, dias = await _get_ultima_asignacion_info(session, entity.id)
        responses.append(_map_to_response(entity, depto, estado_asig, dias))
    return responses


async def count(
    session: AsyncSession,
    filters: dict[str, Any],
) -> int:
    """Cuenta PQRSDF con los filtros aplicados."""
    query = select(func.count(Pqrsdf.id))

    if filters.get("estado"):
        query = query.where(Pqrsdf.estado == filters["estado"])
    if filters.get("tipo"):
        query = query.where(Pqrsdf.tipo == filters["tipo"])
    if filters.get("numeroDocumento"):
        query = query.where(Pqrsdf.numero_documento == filters["numeroDocumento"])
    if filters.get("fechaDesde"):
        query = query.where(Pqrsdf.fecha_radicado >= filters["fechaDesde"])
    if filters.get("fechaHasta"):
        query = query.where(Pqrsdf.fecha_radicado <= filters["fechaHasta"])
    if filters.get("consecutivo"):
        query = query.where(Pqrsdf.consecutivo.ilike(f'%{filters["consecutivo"]}%'))

    result = await session.execute(query)
    return result.scalar() or 0


async def update(
    session: AsyncSession,
    id: int,
    req: UpdatePqrsdfRequest,
    employee_id: int,
) -> PqrsdfResponse | None:
    """Actualiza clasificación y observaciones de una PQRSDF."""
    entity = await pqrsdf_repo.find_by_id(session, id)
    if entity is None:
        return None

    if req.clasificacion is not None:
        entity.clasificacion = req.clasificacion
    if req.observaciones is not None:
        entity.observaciones = req.observaciones

    entity.updated_by = req.updatedBy or employee_id
    entity.updated_at = datetime.now(timezone.utc)

    await pqrsdf_repo.save(session, entity)

    await _registrar_trazabilidad(
        session,
        id,
        "ACTUALIZACION",
        "PQRSDF actualizada",
        employee_id,
    )

    depto, estado_asig, dias = await _get_ultima_asignacion_info(session, id)
    return _map_to_response(entity, depto, estado_asig, dias)


async def responder(
    session: AsyncSession,
    id: int,
    req: ResponderPqrsdfRequest,
    employee_id: int,
) -> PqrsdfResponse | None:
    """Responde una PQRSDF: actualiza respuesta, cambia estado a RESPONDIDO."""
    entity = await pqrsdf_repo.find_by_id(session, id)
    if entity is None:
        return None

    entity.respuesta_final = req.respuestaFinal
    entity.medio_respuesta = req.medioRespuesta
    entity.fecha_respuesta = date.today()
    entity.estado = "RESPONDIDO"
    if req.clasificacion is not None:
        entity.clasificacion = req.clasificacion
    entity.updated_by = req.updatedBy or employee_id
    entity.updated_at = datetime.now(timezone.utc)

    await pqrsdf_repo.save(session, entity)

    # Marcar asignaciones activas como RESPONDIDA
    asignaciones = await asignacion_repo.find_by_pqrsdf_id(session, id)
    for asig in asignaciones:
        if asig.estado in ("PENDIENTE", "EN_PROCESO"):
            asig.estado = "RESPONDIDA"
            asig.fecha_respuesta_area = datetime.now(timezone.utc)
            await asignacion_repo.save(session, asig)

    await _registrar_trazabilidad(
        session,
        id,
        "RESPUESTA",
        f"PQRSDF respondida vía {req.medioRespuesta}",
        employee_id,
    )

    depto, estado_asig, dias = await _get_ultima_asignacion_info(session, id)
    return _map_to_response(entity, depto, estado_asig, dias)


async def cerrar(
    session: AsyncSession,
    id: int,
    employee_id: int,
) -> PqrsdfResponse | None:
    """Cierra una PQRSDF cambiando estado a CERRADO."""
    entity = await pqrsdf_repo.find_by_id(session, id)
    if entity is None:
        return None

    entity.estado = "CERRADO"
    entity.updated_by = employee_id
    entity.updated_at = datetime.now(timezone.utc)

    await pqrsdf_repo.save(session, entity)

    await _registrar_trazabilidad(
        session,
        id,
        "CIERRE",
        "PQRSDF cerrada",
        employee_id,
    )

    depto, estado_asig, dias = await _get_ultima_asignacion_info(session, id)
    return _map_to_response(entity, depto, estado_asig, dias)


async def get_trazabilidad(
    session: AsyncSession,
    pqrsdf_id: int,
) -> list[TrazabilidadResponse]:
    """Obtiene el historial de trazabilidad de una PQRSDF."""
    trazabilidad_list = await trazabilidad_repo.find_by_pqrsdf_id(session, pqrsdf_id)
    return [
        TrazabilidadResponse(
            id=t.id,
            pqrsdfId=t.pqrsdf_id,
            accion=t.accion,
            descripcion=t.descripcion,
            usuarioId=t.usuario_id,
            usuarioNombre=t.usuario_nombre,
            metadataJson=t.metadata_json,
            createdAt=t.created_at,
        )
        for t in sorted(trazabilidad_list, key=lambda x: x.created_at or datetime.min)
    ]
